ユーザー管理機能

mokamokononono opened the pull request:
# What
+ ユーザー管理機能の実装
  + deviseの導入
  + 新規登録
  + ログイン/ログアウト
  + gyazo
    + 新規登録/ログアウト https://gyazo.com/cde90435e4d675fe1618baebaaebdeb3
    + ログイン/ログアウト https://gyazo.com/e8391e2c722aac0a937353b8d9976ade
    + 新登録失敗 https://gyazo.com/c59c6488672d7113a2d002177af189db
    + ログイン失敗 https://gyazo.com/dcb1d44f127eba81b6fd7852017f60ab
+ CSSファイルの追加
+ Userモデル単体テストコードの記述
  + rspecとその周辺のgemの導入
  + ログイン時の単体テストコード
  + https://gyazo.com/014d958910bb0d66c7b0af72a210845d

# Why
+ ユーザー管理機能に必須な新規登録/ログイン/ログアウトができるまで実装する
+ モデルの単体テストを行い、正常にユーザー管理機能が動作するか確認する

mokamokononono committed:
master commit

mokamokononono committed:
ユーザー管理機能の実装

・新規登録
・ログイン/ログアウト
・CSSファイルの追加

mokamokononono committed:
Userモデルのバリデーション設定

・空データの禁止
・ニックネームの一意性制約

mokamokononono committed:
rspec単体テストの導入

mokamokononono committed:
Userモデル単体テストコードの記述

wakita0217 commented:
こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？

```
・新規登録に成功する動画
・新規登録に失敗し、エラーが表示される動画
・ログインに成功する動画
・ログインに失敗し、エラーが表示される動画
・ログアウトに成功する動画
```

生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

お手数ですが、何卒よろしくお願い致します。

mokamokononono committed:
新規登録/ログイン画面のエラーメッセージ表示修正

mokamokononono committed:
微細な修正

mokamokononono commented:
> こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？
> 
> ```
> ・新規登録に成功する動画
> ・新規登録に失敗し、エラーが表示される動画
> ・ログインに成功する動画
> ・ログインに失敗し、エラーが表示される動画
> ・ログアウトに成功する動画
> ```
> 
> 生成されたURLをこのPRに貼っていただければ幸いです。
> その後、再度レビュー依頼フォームにてご提出をお願いいたします。
> 
> お手数ですが、何卒よろしくお願い致します。

# what
上記を対応しました

tatsuki-kato commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,15 @@
+  protected
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(
+      :sign_up, keys: [:nick_name, :email, :kanji_family_name, :kanji_given_name, :katakana_family_name, :katakana_given_name, :date_of_birth]
```
```rb
      :sign_up, keys: [:nick_name, :email, :kanji_family_name, :kanji_given_name, :katakana_family_name, :katakana_given_name, :date_of_birth]
```
上記の中の`:email`は記述がなくても問題ないかと思いますので、削除しましょう。
理由は、:sign_upの中でemailを取得するようにしているためです。

tatsuki-kato commented on app/models/user.rb:
```diff
@@ -0,0 +1,15 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  validates :nick_name, presence: true
+  validates :kanji_family_name, presence: true
+  validates :kanji_given_name, presence: true
+  validates :katakana_family_name, presence: true
+  validates :katakana_given_name, presence: true
+  validates :date_of_birth, presence: true
```
バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となりますので、`presence: ture`については、with_optionsでまとめておきましょう！

tatsuki-kato commented on app/models/user.rb:
```diff
@@ -0,0 +1,15 @@
+  validates :katakana_family_name, presence: true
+  validates :katakana_given_name, presence: true
+  validates :date_of_birth, presence: true
+
+  validates :nick_name, uniqueness: true
+end
```
以下のバリデーションを追加しましょう。
```
- ユーザー本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
- ユーザー本名のフリガナは、全角（カタカナ）での入力が必須であること
- パスワードは、半角英数字混合での入力が必須であること（半角英数字が混合されていれば、登録が可能なこと）
```

tatsuki-kato commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+<%= render "shared/second-header"%>
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
```erb
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
上記のようなコメントアウトは実装方針の内容であり、企業にポートフォリオとして提出する際に、好ましくないため削除しましょう！
他の部分も実装済みであれば削除しましょう。

tatsuki-kato commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,77 @@
+        @user.password_confirmation = '12345'
+        @user.valid?
+        expect(@user.errors.full_messages).to include "Password is too short (minimum is 6 characters)"
+      end
+    end
+  end
+end
```
バリデーションを追加したあと、以下のテストも追加しましょう。
```
名字は漢字・ひらがな・カタカナ以外では登録できない
名前はは漢字・ひらがな・カタカナ以外では登録できない
名字（ふりがな）はカタカナ以外では登録できない
名前（ふりがな）はカタカナ以外では登録できない
パスワードは英字のみでは登録できない
パスワードは数字のみでは登録できない
パスワードは全角では登録できない
```

tatsuki-kato reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認をお願いします！

mokamokononono commented:
# What
+ Userモデルのバリデーション追加/修正
  + with_optionsを用いた記述に変更
  + 全角苗字と名前、全角カタカナの苗字と名前のvalidation
  + passwordの半角英数字混合のvalidation
  + email validationの削除（重複）
+ Userモデル単体テストコードの修正
  + gem 'gimei'の導入
    + factory botで全角漢字かなカナ文字に対応するため
    + カリキュラム外内容（個人で調べた上での実装）
    +  Gemfile, spec/factories/users.rbの変更
  + 単体テストコードの追加
    + 全角苗字と名前、全角カタカナの苗字と名前のvalidationのチェック
    + passwordの半角英数字混合のvalidationチェック　
  + https://gyazo.com/c573495a2299fa5e0beec886c97e8407

# Why
+ コードレビュー結果から修正し、ユーザー管理機能に必須な新規登録/ログイン/ログアウトができるまで実装する
+ コードレビュー結果から修正し、モデルの単体テストを行い、正常にユーザー管理機能が動作するか確認する

mokamokononono committed:
バリデーションと単体テストコードの修正

・全角文字に対応したバリデーション
・半角英数字混合に対応したパスワードのバリデーション
・gem 'gimei'の導入
・全角文字に対応したテストコードとfactory bot
・半角英数字混合に対応したパスワードのテストコード

mokamokononono commented:
# What
+ Userモデルのバリデーション追加/修正
  + with_optionsを用いた記述に変更
  + 全角苗字と名前、全角カタカナの苗字と名前のvalidation
  + passwordの半角英数字混合のvalidation
  + email validationの削除（重複）
+ Userモデル単体テストコードの修正
  + gem 'gimei'の導入
    + factory botで全角漢字かなカナ文字に対応するため
    + カリキュラム外内容（個人で調べた上での実装）
    +  Gemfile, spec/factories/users.rbの変更
  + 単体テストコードの追加
    + 全角苗字と名前、全角カタカナの苗字と名前のvalidationのチェック
    + passwordの半角英数字混合のvalidationチェック　
  + https://gyazo.com/c573495a2299fa5e0beec886c97e8407

# Why
+ コードレビュー結果から修正し、ユーザー管理機能に必須な新規登録/ログイン/ログアウトができるまで実装する
+ コードレビュー結果から修正し、モデルの単体テストを行い、正常にユーザー管理機能が動作するか確認する

d-kazu4k commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,115 @@
+        another_user.valid?
+        expect(another_user.errors.full_messages).to include "Nick name has already been taken"
+      end
+      it 'passwordが5文字以下では登録できない' do
+        @user.password = '12345'
+        @user.password_confirmation = '12345'
+        @user.valid?
+        expect(@user.errors.full_messages).to include "Password is too short (minimum is 6 characters)"
+      end
```
```ruby
      it 'passwordが5文字以下では登録できない' do
        @user.password = '12345'
        @user.password_confirmation = '12345'
        @user.valid?
        expect(@user.errors.full_messages).to include "Password is too short (minimum is 6 characters)"
      end
```
代入する値を半角英数混合の5文字にしましょう！

d-kazu4k commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+    katakana_given_name   {Gimei.first.katakana}
+    date_of_birth         {Faker::Date.birthday(min_age: 18, max_age: 65)}
+    email                 {Faker::Internet.free_email}
+    password              {Faker::Internet.password(min_length: 6)}
```
```ruby
password {Faker::Internet.password(min_length: 6)}
```
上記ですが、半角英数混合で値が作られない可能性があるので、必ず半角英数混合になるような記述を代入しましょう！

下記を試していただければと存じます。
```ruby
password {"0a" + Faker::Internet.password(min_length: 6)}
```

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

mokamokononono commented:
# What
- 単体テストコードの修正
  - パスワード入力の半角英数字混合入力対応 
    -  it 'passwordが5文字以下では登録できない' do
    -  factory bot の password
  - https://gyazo.com/1355c73c990df16627236039fd71d55a

# Why
+ コードレビュー結果から修正し、ユーザー管理機能に必須な新規登録/ログイン/ログアウトができるまで実装する
+ コードレビュー結果から修正し、モデルの単体テストを行い、正常にユーザー管理機能が動作するか確認する

mokamokononono commented:
# What
- 単体テストコードの修正
  - パスワード入力の半角英数字混合入力対応 
    -  it 'passwordが5文字以下では登録できない' do
    -  factory bot の password
  - https://gyazo.com/1355c73c990df16627236039fd71d55a

# Why
+ コードレビュー結果から修正し、ユーザー管理機能に必須な新規登録/ログイン/ログアウトができるまで実装する
+ コードレビュー結果から修正し、モデルの単体テストを行い、正常にユーザー管理機能が動作するか確認する
